streams/src/main/java/org/apache/kafka/streams/state/internals/StoreChangeLogger.java

```diff
@@ -109,7 +109,7 @@ public void logChange(ValueGetter<K, V> getter) {
             }
             for (K k : this.dirty) {
                 V v = getter.get(k);
-                collector.send(new ProducerRecord<>(this.topic, this.partition, k, v), keySerializer, valueSerializer);
+                collector.send(new ProducerRecord<>(this.topic, this.partition, context.timestamp(), k, v), keySerializer, valueSerializer);
```

Comment:
One thought about log rolling: from KIP-33 we have one known issue of timestamp based log rolling, quote:

```
Enforce time based log rolling

Currently time based log rolling is based on the creating time of the log segment. With this KIP, the time based rolling would be changed to based on the timestamp of the first message in the log segment. A new log segment will be rolled out if current time is greater than the timestamp of the first message in the segment + log.roll.ms. When message.timestamp.type=CreateTime, user should set max.message.time.difference.ms appropriately together with log.roll.ms to avoid frequent log segment roll out.
```

In a word, Kafka's log rolling assumes that the timestamp specified are somewhat related to real time system clocks, while user-defined timestamp "unit" in Kafka Streams can have arbitrary semantics; if a user specified timestamp "jumps" in large span, it could potentially cause the log to roll upon each append to the changelog.

So this leads me to think whether or not should we define `window retention` based on system time, not on event time, and hence produce changelog messages still with no timestamp and let server to use `LogAppendTime` to assign the timestamps. Thoughts @dguy ?